### ISSUE TITLE
High iowait after backups on slow RAID-1 volumes

### ISSUE THREAD

**Issue by kakduman** [pterodactyl/panel#2946]
**Background (please complete the following information):**
* Panel or Daemon: Wings
* Version of Panel/Daemon: Wings 1.2.0
* Server's OS: Debian 10.5
* Your Computer's OS & Browser: Windows 10, Google Chrome
* Relevant server hardware: 3900XT, storage is NVMe in RAID-1.

**Describe the bug**
Creating an archive of a large file or folder (15GB+) will cause iowait to spiral out of control. iowait stays high for a few minutes even after the archive is created for about 5 minutes. Wings initially peaks in iotop, but it seems like the main issue is with RAID. iotop shows that maintaining RAID-1 for the archive creation is using high io, even after the archive is created. If it was strictly related to RAID, I would expect this issue to occur for the tar command as well, but it only results in high iowait after creating an archive through Pterodactyl's system. During this ~5 minute period of high iowait, servers running on the node have their CPU usage throttled to just ~2% or so, and a few crash. Anyone is able to cause the lag spike simply by creating an archive of a large folder such as their world folder.
![image](https://user-images.githubusercontent.com/43528123/103430827-b12e3700-4b8d-11eb-9642-48af3a78c6e3.png)
You can view the netdata output for this archive creation here: [https://drive.google.com/file/d/1M9cS29ch8XPe65rkKEhMAhseWJz7YOq8/view?usp=sharing](https://drive.google.com/file/d/1M9cS29ch8XPe65rkKEhMAhseWJz7YOq8/view?usp=sharing)

**To Reproduce**
I'm not sure... This isn't as clear-cut as the issue where backups were using a ton of iowait when backing up to a slow device.
I have four nodes: three with NVMe storage in RAID-1, and one with NVMe storage not in RAID. I can only reliably reproduce the issue on one node: the most active one which has the greatest amount of memory usage. I'm not sure if this is related, but it might be possible that the node already needs to have a large portion of memory in use in order to reproduce the issue(?)

It only occurs when creating a .tar.gz archive through Pterodactyl, which is why I believe it is an issue with Pterodactyl and not the node. Typing the command `tar vzcf file.tar.gz file.txt` does not result in high iowait. For the node that the bug occurs on, I'm able to reproduce it with the following steps.:
1. Create a file with `< /dev/urandom tr -dc "[:alnum:]" | head -c15000000000 > file.txt`.
2. Use the panel to add this file to an archive.
3. Observe high iowait even after the archive is created.

**Expected behavior**
Similar to the tar command or backups, I would not expect creating an archive to use so high iowait even after the archive is created. Backups work fine, and so does the tar command.

**Comment by matthewpi** [comment]
The backup system and File Manager archive endpoint use basically the same internal logic except for when walking files as the backup system uses an ignore file while the endpoint uses an array of file names.  I find it hard to believe that this is caused by Pterodactyl if wings is not the one showing up in iotop.

**Comment by kakduman** [comment]
Sorry, I probably should have clarified: wings is the one showing up in iotop during the archive. It later gets replaced with raid1 after the archive. When archiving through the tar command, there is no noticeable uptick in iowait from raid1 (nor from anything else), leading me to believe that something wings is doing is causing raid1 to have high iowait after the archive creation. Meanwhile, wings itself causes high iowait when starting and during the archive creation.

Here is the iowait when starting (and during) an archive:
![image](https://user-images.githubusercontent.com/43528123/103431419-0457b800-4b95-11eb-92bb-1d20002210aa.png)

I understand if the issue can't be resolved because I can't figure out how to reproduce it on all systems. Please let me know what information I might be able to provide or what steps I might be able to try in order to help resolve the issue, if you think it is doable.

**Comment by kakduman** [comment]
After some further testing it appears that this happens with backups too. The reason it wasn't happening earlier is that my backup directory is mounted to my HDD whereas archiving files takes place on the NVMe RAID-1 volume. After switching the backup directory to the RAID-1 volume, similar iowait spikes can be observed.

It also appears that the vast majority of the iowait spikes occur AFTER the archive rather than before it due to `md2_raid1` rather than directly from wings.

However, it's interesting that the high iowait caused by `md2_raid1` is not observed when using the tar command instead of backing up within pterodactyl. It's also interesting that wings creates backups significantly faster than the tar command. Is it possible that wings dumps data so fast that the drives have trouble synchronizing the data between them, causing high iowait after wings finishes writing the data?

I tried throttling wings' write speed to 35 M/s which should make it easier for the drives to synchronize after the archive is created, and that prevented all lag after the archive was created. iowait still peaked at ~35% during the archive, but there was no uptick in iowait after the archive was finished completing.

I believe the reason that the issue only occurs on one node is that that node has NVMe drives which are slower than those on the other nodes. The node that the issue occurs on has Crucial P1 drives.

Using this knowledge, the following reproduction steps can be created:
1. Create a RAID-1 volume on slow drives, and make sure that servers are located on this volume.
2. Create a file with `< /dev/urandom tr -dc "[:alnum:]" | head -c15000000000 > file.txt`
3. Use the panel to add this file to an archive.
4. Observe high iowait even after the archive is created.

Even though my NVMe drives aren't the fastest, they're not *slow* either. I think many users may have slower drives in RAID-1, so wings should consider incorporating a better way to handle the "slow" RAID-1 volumes. I think potential ways to alleviate the issue could include:
* Separate write limit throttles for backups and archives. This would patch the issue for people like me who have backups and archives in different volumes, and it wouldn't affect other users at all.
* Probably infeasible: rather than archiving within the server's directory, archive in a new archive directory specified in the config.yml. Then, create a symlink to that archive in the server's directory. When unarchiving files, transfer them from the archives directory to the server's directory. This would completely resolve the issue for people like me, but would still require other users to specify throttling speeds. It's also overcomplicated and probably has a bunch of oversights that I didn't think of.
* Possibly something else? I'm not sure. Ideally something that wouldn't require the user to be aware that they need to throttle their archive speeds.

**Comment by matthewpi** [comment]
> Separate write limit throttles for backups and archives.

I don't see how this really solves the problem, I have tested backups with no write limit (my old drive hit 200 MiB/s when archiving) and the iowait occasionally spiked to like 5% but mainly stayed at < 1-2%.  So adding a limit is more so avoiding the problem that you are actually facing rather than finding an actual solution.

> rather than archiving within the server's directory, archive in a new archive directory specified in the config.yml.

This in my opinion is a hacky workaround for a problem that should have other solutions.  Not to mention, the filesystem intentionally blocks symlinks outside of the root directory of a server and changing this logic could become a security risk.

I wonder if this is related to using a large single file or potentially the issue of applications being forced to swap due to the amount of memory usage, this may be apart of a larger problem but from what I can currently tell, our logic in no way should mess with RAID or cause that amount of I/O usage.

EDIT 1: You may also want to actually check the health of your disk and do more research on I/O usage from `mdX_raid1` (specifically referencing this post on [StackOverflow](https://serverfault.com/questions/391425/high-load-of-resources-by-software-raid1-in-debian-6) even though it is old as hell).  You may also want to check your swap as well and maybe try disabling it.

EDIT 2: This also could be related to the disk usage checker going hard on the server, but if you have very few files it shouldn't be too bad.

**Comment by kakduman** [comment]
The error also occurs when archiving a folder with lots of smaller files rather than just one large file. I talked with the Syscraft Discord and the error may be related to the QLC drive's horrible 4k performance rather than an issue with drive health. I'll probably need to upgrade that. Closing the issue since I'm hoping the majority of people won't use Crucial P1 drives.

Thanks for all the help, you're the best ❤️
